File: ParkSeongGeun/#2 Implementation/2-7.cpp
#include <bits/stdc++.h>

using namespace std;

string s;
int ans=1000; // 입력 될 수 있는 최대 문자열 길이 == MAX_VALUE

int solution(string s) {
	for (int i = 1; i <= s.size()/2; i++) { // 절반까지만 비교, 절반 이상 길이를 비교하는 건 의미 X
		int cnt = 1; // 몇 번 반복되었는 지 확인
		string res = ""; // 최종 변화된 문자열
		string ss = s.substr(0, i); // abcabcd 가 있을 떄 0번째부터 i개만큼 추출

		for (int j = i; j <= s.size(); j+=i) { // 그 뒤부터 검사 시작
			// i개 만큼 같으면 cnt 증가
			if (ss == s.substr(j, i)) {
				cnt += 1;
			}
			// 다를 때 cnt가 1이면 한번만 반복된 것 -> 그냥 res에 추출한 것만 더해주기
			else {
				if (cnt == 1) {
					res += ss;
				}
				// 1이 아니면 반복된 개수 cnt + ss를 res에 붙여줌
				else {
					res += (to_string(cnt)) + ss;
				}
				// 그 후 그 자리부터 i개만큼을 새로운 기준으로 삼기
				ss = s.substr(j, i);
				cnt = 1;
			}
		}
		// s.size() 가 i개씩 안 끊어지는 경우 -> 맨 끝에 남은 부분은 그냥 res에 더하기
		if ((s.size() % i) != 0) {
			res += s.substr((s.size() / i) * i);
		}
		// ans와 res의 길이 비교
		if (ans > res.size()) {
			ans = res.size();
		}
	}
	// ans 반환
	return ans;
}

int main() {
	cin >> s;
	cout << solution(s)<<"\n";
	return 0;
}
// aabbaccc
// ababcdcdababcdcd
// abcabcdede
// abcabcabcabcdededededede

File: yoonezi/#2 Implementation/2-7.py
# 문자열 + 구현 문제

# - 첫번째부터 문자열의 절반까지만 for문으로 순회하며 압출할 길이 결정
# - 압축할 내용을 담고 있다가 한번에 숫자 + 문자 형태로 만들기
# * 압출 할 수 없을 땐 숫자없이 문자만 있도록
# - 압축 단위마다 최소값을 갱신


def solution(s):
    answer = len(s)
    # 입력된 문자열의 개수가 하나일 경우 처리
    if len(s) == 1:
        return 1
        
    # 반복되는 문자열은 최대 문자열의 반이니
    for i in range(1, len(s) // 2 + 1):
        # print("i:", i)
        cnt = 1 
        arr = '' # 문자 추가할 배열
        tmp = s[:i]
        # print('tmp:', tmp)
        for j in range(i, len(s), i):
            # print(i, len(s), i)
            # print('j:', j)
            if tmp == s[j:j+i]:
                cnt += 1
                # print(cnt)
                # print(tmp)
            else:
                if cnt == 1: # 압축 할 수 없는 경우
                    arr += tmp
                    # print(arr)
                else: # 압축 할 수 있는 경우
                    arr += str(cnt) + tmp
                    # print(arr)
                cnt = 1
                tmp = s[j:j+i] # 여기서 부터 다시 시작
                # print(tmp)
                
        if cnt == 1: # 남은 문자열 처리
            arr += tmp
            # print(arr)
        else:
            arr += str(cnt) + tmp
            # print(arr)
        answer = min(answer, len(arr)) # 계속 갱신하며 최소값 찾기
        # print(answer)
                    
    return answer


print(solution('aabbaccc'))
# print(solution('ababcdcdababcdcd'))
# print(solution('abcabcdede'))
# print(solution('abcabcabcabcdededededede'))
